### PR TITLE
fonts: Add support for the CSS font matching algorithm

### LINKED ISSUE #20684: Intermittent results in /css/css-fonts/variations/at-font-face-font-matching.html
```
  ▶ Unexpected subtest result in /css/css-fonts/variations/at-font-face-font-matching.html:
  │ FAIL [expected PASS] Matching font-style: 'oblique 20deg' should prefer 'oblique 30deg 60deg' over 'oblique 40deg 50deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 20deg' should prefer 'oblique 40deg 50deg' over 'oblique 10deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 20deg' should prefer 'oblique 10deg' over 'italic'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 20deg' should prefer 'italic' over 'oblique 0deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 20deg' should prefer 'oblique 0deg' over 'oblique -50deg -20deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 20deg' should prefer 'oblique -50deg -20deg' over 'oblique -40deg -30deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 21deg' should prefer 'oblique 21deg' over 'oblique 30deg 60deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ FAIL [expected PASS] Matching font-style: 'oblique 21deg' should prefer 'oblique 30deg 60deg' over 'oblique 40deg 50deg'
  │   → assert_true: Unexpected font on test element expected true got false
  │ 
  │ verifyFont@http://web-platform.test:8000/css/css-fonts/variations/at-font-face-font-matching.html:72:13
  └ testDescriptor/</</<@http://web-platform.test:8000/css/css-fonts/variations/at-font-face-font-matching.html:110:33
```

### LINKED ISSUE #20686: Intermittent results in /css/css-fonts/variations/font-weight-matching.html
```
  ▶ Unexpected subtest result in /css/css-fonts/variations/font-weight-matching.html:
  └ PASS [expected FAIL] Test @font-face matching for weight 470

  ▶ Unexpected subtest result in /css/css-fonts/variations/font-weight-matching.html:
  │ FAIL [expected PASS] Test @font-face matching for weight 99
  │   → assert_approx_equals: @font-face should be mapped to CSSTest Weights 900. expected 90 +/- 2 but got 180
  │ FAIL [expected PASS] Test @font-face matching for weight 100
  │   → assert_approx_equals: @font-face should be mapped to CSSTest Weights 900. expected 90 +/- 2 but got 180
  │ FAIL [expected PASS] Test @font-face matching for weight 249
  │   → assert_approx_equals: @font-face should be mapped to CSSTest Weights 900. expected 90 +/- 2 but got 180
  │ 
  └ @http://web-platform.test:8000/css/css-fonts/variations/font-weight-matching.html:128:65
```

### PR DESCRIPTION
**Pull request by mrobinson**
This is a port from Gecko of the CSS font matching algorithm distance
functions as well as the "simple family" concept for optimizing
matching when dealing with simple fonts.

Fixes #189.
Fixes #190.
Fixes #20686.
Fixes #20684.

Co-authored-by: Mukilan Thiyagarajan <mukilan@igalia.com>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes

### CODE REVIEW

File: components/gfx/font_template.rs [hunk 1]
```diff
@@ -67,6 +59,12 @@ impl FontTemplateDescriptor {
...
     }
 
+    pub fn is_variation_font(&self) -> bool {
+        self.weight.0 != self.weight.1 ||
+            self.stretch.0 != self.stretch.1 ||
+            self.style.0 != self.style.1
+    }
+
     /// Returns a score indicating how far apart visually the two font descriptors are. This is
     /// used for fuzzy font selection.
```
Comment (mukilan) [inline components/gfx/font_template.rs:69]: Should we maybe replace `fuzzy font selection` with CSS font matching?

File: components/gfx/font_template.rs [hunk 2]
```diff
@@ -75,19 +73,29 @@ impl FontTemplateDescriptor {
...
+        // Sanity-check that the distances are within the expected range
+        // (update if implementation of the distance functions is changed).
+        assert!(stretch_distance >= 0.0 && stretch_distance <= 2000.0);
+        assert!(style_distance >= 0.0 && style_distance <= 500.0);
+        assert!(weight_distance >= 0.0 && weight_distance <= 1600.0);
+
+        // Factors used to weight the distances between the available and target font
+        // properties during font-matching. These ensure that we respect the CSS-fonts
+        // requirement that font-stretch >> font-style >> font-weight; and in addition,
+        // a mismatch between the desired and actual glyph presentation (emoji vs text)
```
Comment (mukilan) [inline components/gfx/font_template.rs:90]: Should we make the emoji/presentation mismatch part into a TODO comment?

File: components/gfx/font_store.rs [hunk 3]
```diff
@@ -132,10 +134,55 @@ impl WebRenderFontStore {
     }
 }
 
-/// A list of font templates that make up a given font family.
+/// A struct that represents the available templates in a "simple family." A simple family
+/// is one that contains <= 4 available faces regular, bold, italic, and bold italic. Having
```
Comment (mukilan) [inline components/gfx/font_store.rs:138]:
```suggestion
/// is one that contains <= 4 available faces: regular, bold, italic, and bold italic. Having
```

File: components/gfx/font_template.rs [hunk 4]
```diff
@@ -286,3 +296,278 @@ impl FontTemplateRefMethods for FontTemplateRef {
...
+            return REVERSE + NEGATE + (min_angle - target_angle);
+        }
+        if min_style == FontStyle::ITALIC {
+            return REVERSE + NEGATE - 2.0;
+        }
+        return REVERSE + NEGATE - 1.0;
+    }
+
+    fn to_float(&self) -> f32 {
+        todo!()
```
Comment (mukilan) [inline components/gfx/font_template.rs:561]: Should we make this `unimplemented!()`?

File: components/gfx/font_store.rs [hunk 5]
```diff
@@ -132,10 +134,55 @@ impl WebRenderFontStore {
     }
 }
 
-/// A list of font templates that make up a given font family.
+/// A struct that represents the available templates in a "simple family." A simple family
+/// is one that contains <= 4 available faces: regular, bold, italic, and bold italic. Having
+/// this simple family abstraction makes font matching much faster for families that don't
+/// have a complex set of  fonts.
```
Comment (mukilan) [inline components/gfx/font_store.rs:140]:
```suggestion
/// have a complex set of fonts.
```

**Review by mukilan — APPROVED** [review]

File: components/gfx/font_store.rs (hunk 5 above)
Comment (mrobinson) [reply components/gfx/font_store.rs]: I've fixed this one.

**Pull request merged**